File: web/src/components/garageDisplay.test.ts
import { describe, expect, it } from "vitest"

import type { GarageMemberDrives, Place, Vehicle } from "@/api/types"
import {
  drivenByLabel,
  groupVehiclesByKeptAt,
  vehicleYearOptions,
} from "@/components/garageDisplay"

function vehicle(partial: Partial<Vehicle> & Pick<Vehicle, "id" | "label">): Vehicle {
  return {
    ownerAdultId: "1",
    driverAdultIds: ["1"],
    keptAtPlaceId: null,
    year: 2019,
    make: "HONDA",
    model: "Civic",
    seats: 5,
    suggestedSeats: null,
    ...partial,
  }
}

describe("vehicleYearOptions", () => {
  it("runs from 1996 through the next UTC calendar year", () => {
    const years = vehicleYearOptions(new Date("2026-08-14T00:00:00Z"))
    expect(years[0]).toBe(2027)
    expect(years.at(-1)).toBe(1996)
    expect(years).toHaveLength(2027 - 1996 + 1)
  })
})

describe("groupVehiclesByKeptAt", () => {
  const mom: Place = {
    id: "p1",
    name: "Mom's house",
    address: "1 Rd",
    latitude: 1,
    longitude: 2,
  }
  const grandma: Place = {
    id: "p2",
    name: "Grandma's house",
    address: "2 Rd",
    latitude: 3,
    longitude: 4,
  }

  it("groups by kept-at place then Other, without sharing by house", () => {
    const vehicles = [
      vehicle({ id: "van", label: "Blue van", keptAtPlaceId: "p1" }),
      vehicle({ id: "camry", label: "Camry", keptAtPlaceId: "p2", ownerAdultId: "g" }),
      vehicle({ id: "truck", label: "Truck", keptAtPlaceId: "p2", ownerAdultId: "gp" }),
      vehicle({ id: "civic", label: "Civic", keptAtPlaceId: null, ownerAdultId: "n" }),
    ]
    const groups = groupVehiclesByKeptAt(vehicles, [mom, grandma])
    expect(groups.map((g) => g.heading)).toEqual(["Mom's house", "Grandma's house", "Other"])
    expect(groups[1]?.vehicles.map((v) => v.label)).toEqual(["Camry", "Truck"])
    expect(groups[2]?.vehicles.map((v) => v.label)).toEqual(["Civic"])
  })
})

describe("drivenByLabel", () => {
  const members: GarageMemberDrives[] = [
    { adultId: "1", displayName: "Mom", drives: true },
    { adultId: "2", displayName: "Dad", drives: true },
  ]

  it("lists driver display names", () => {
    expect(
      drivenByLabel(
        vehicle({ id: "v", label: "Van", driverAdultIds: ["1", "2"] }),
        members,
      ),
    ).toBe("Driven by Mom, Dad")
  })
})

File: mobile/iosApp/scripts/garage_display_test.swift
import Foundation

@main
struct GarageDisplayTestMain {
    static func main() {
        func expect(_ condition: Bool, _ message: String) {
            if !condition {
                fputs("FAIL: \(message)\n", stderr)
                exit(1)
            }
        }

        let years = GarageDisplay.yearOptions(
            now: ISO8601DateFormatter().date(from: "2026-08-14T00:00:00Z")!
        )
        expect(years.first == 2027, "year options start at next UTC year")
        expect(years.last == 1996, "year options end at 1996")

        func vehicle(_ id: String, _ label: String, keptAt: String?, owner: String = "1") -> GarageVehicleView {
            GarageVehicleView(
                id: id,
                ownerAdultId: owner,
                driverAdultIds: [owner],
                keptAtPlaceId: keptAt,
                label: label,
                year: 2019,
                make: "HONDA",
                model: "Civic",
                seats: 5,
                suggestedSeats: nil
            )
        }
        let groups = GarageDisplay.groupVehicles(
            [
                vehicle("van", "Blue van", keptAt: "p1"),
                vehicle("camry", "Camry", keptAt: "p2", owner: "g"),
                vehicle("truck", "Truck", keptAt: "p2", owner: "gp"),
                vehicle("civic", "Civic", keptAt: nil, owner: "n"),
            ],
            places: [("p1", "Mom's house"), ("p2", "Grandma's house")]
        )
        expect(
            groups.map(\.heading) == ["Mom's house", "Grandma's house", "Other"],
            "groups by kept-at then Other"
        )
        expect(groups[1].vehicles.map(\.label) == ["Camry", "Truck"], "same house does not merge owners")
        expect(
            GarageDisplay.drivenByLabel(
                vehicle("v", "Van", keptAt: nil),
                members: [
                    GarageMemberView(adultId: "1", displayName: "Mom", drives: true),
                ]
            ) == "Driven by Mom",
            "driven by uses display names"
        )

        let json = """
        {"members":[{"adultId":"1","displayName":"Mom","drives":true}],"vehicles":[]}
        """
        expect(GarageView.decode(json)?.members.first?.drives == true, "decodes garage JSON")

        let contentURL = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .appendingPathComponent("iosApp/ContentView.swift")
        let content = try! String(contentsOf: contentURL, encoding: .utf8)
        expect(content.contains("title: \"Garage\""), "More has Garage row")
        expect(content.contains("I don’t drive"), "Garage has don't-drive toggle")
        expect(content.contains("Add vehicle"), "Garage has Add vehicle")
        expect(content.contains("Who can drive this?"), "Garage has who-can-drive")
        expect(content.contains("Seats (including driver)"), "Garage seats include driver")
        expect(!content.contains("VIN"), "Garage has no VIN field")
        expect(content.contains("case .garage:"), "More navigation includes garage")

        print("GarageDisplay tests passed")
    }
}
